File: samples/javascript/session-security.js
/**
 * MITRE ATT&CK T1185: Browser Session Hijacking - Defense Implementation
 * Educational demonstration of session security and hijacking detection
 * 
 * @description Shows vulnerable vs secure session management in Express.js
 * @reference https://attack.mitre.org/techniques/T1185/
 */

const crypto = require('crypto');

// =============================================================================
// SESSION STORAGE (In-Memory for Demo)
// =============================================================================

/**
 * In-memory session store
 * In production, use Redis, Memcached, or encrypted database storage
 */
class SessionStore {
  constructor() {
    this.sessions = new Map();
    this.userSessions = new Map(); // Track sessions per user
  }

  /**
   * Create a new session
   * @param {string} sessionId - Unique session identifier
   * @param {object} data - Session data
   */
  create(sessionId, data) {
    this.sessions.set(sessionId, {
      ...data,
      createdAt: Date.now(),
      lastActivity: Date.now()
    });

    // Track user's sessions for concurrent session limits
    const userId = data.userId;
    if (!this.userSessions.has(userId)) {
      this.userSessions.set(userId, new Set());
    }
    this.userSessions.get(userId).add(sessionId);
  }

  /**
   * Get session data
   * @param {string} sessionId - Session identifier
   * @returns {object|null} Session data or null
   */
  get(sessionId) {
    return this.sessions.get(sessionId) || null;
  }

  /**
   * Update session data
   * @param {string} sessionId - Session identifier
   * @param {object} updates - Data to update
   */
  update(sessionId, updates) {
    const session = this.sessions.get(sessionId);
    if (session) {
      this.sessions.set(sessionId, { ...session, ...updates, lastActivity: Date.now() });
    }
  }

  /**
   * Delete a session
   * @param {string} sessionId - Session identifier
   */
  delete(sessionId) {
    const session = this.sessions.get(sessionId);
    if (session) {
      // Remove from user's session list
      const userSessionSet = this.userSessions.get(session.userId);
      if (userSessionSet) {
        userSessionSet.delete(sessionId);
      }
      this.sessions.delete(sessionId);
    }
  }

  /**
   * Get all sessions for a user
   * @param {string} userId - User identifier
   * @returns {Array} Array of session IDs
   */
  getUserSessions(userId) {
    return Array.from(this.userSessions.get(userId) || []);
  }

  /**
   * Delete all sessions for a user
   * @param {string} userId - User identifier
   */
  deleteUserSessions(userId) {
    const sessionIds = this.getUserSessions(userId);
    sessionIds.forEach(sid => this.sessions.delete(sid));
    this.userSessions.delete(userId);
  }
}

// =============================================================================
// VULNERABLE SESSION MANAGEMENT
// =============================================================================

/**
 * VULNERABLE: Basic session management without security controls
 * ATT&CK T1185 - Browser Session Hijacking
 * 
 * @description Vulnerable to session hijacking, fixation, and replay attacks
 * @vulnerability No fingerprinting, no rotation, no hijack detection
 */
class VulnerableSessionManager {
  constructor() {
    this.store = new SessionStore();
  }

  /**
   * VULNERABLE: Creates session without security controls
   * @param {object} req - Request object
   * @param {string} userId - User identifier
   * @returns {string} Session ID
   */
  createSession(req, userId) {
    // VULNERABLE: Predictable session ID (just a counter or timestamp)
    const sessionId = `session_${Date.now()}_${Math.random()}`;
    
    console.log('🔴 VULNERABLE: Creating session without security controls');
    
    this.store.create(sessionId, {
      userId,
      // VULNERABLE: No fingerprinting data stored
    });

    return sessionId;
  }

  /**
   * VULNERABLE: Validates session without security checks
   * @param {object} req - Request object
   * @param {string} sessionId - Session ID
   * @returns {object|null} Session data
   */
  validateSession(req, sessionId) {
    const session = this.store.get(sessionId);
    
    if (!session) {
      return null;
    }

    // VULNERABLE: No fingerprinting validation
    // VULNERABLE: No hijacking detection
    // VULNERABLE: Sessions never expire or rotate
    
    console.log('🔴 VULNERABLE: Session validated without security checks');
    
    return session;
  }
}

// =============================================================================
// SECURE SESSION MANAGEMENT
// =============================================================================

/**
 * SECURE: Session management with anti-hijacking controls
 * Defense against ATT&CK T1185
 * 
 * @description Implements fingerprinting, rotation, hijack detection, and session limits
 * @security Multiple layers of session security
 */
class SecureSessionManager {
  constructor(options = {}) {
    this.store = new SessionStore();
    this.config = {
      maxConcurrentSessions: options.maxConcurrentSessions || 3,
      sessionTimeout: options.sessionTimeout || 30 * 60 * 1000, // 30 minutes
      rotationInterval: options.rotationInterval || 15 * 60 * 1000, // 15 minutes
      fingerprintStrength: options.fingerprintStrength || 'medium', // low, medium, high
      ...options
    };
  }

  /**
   * SECURE: Creates session with security controls
   * @param {object} req - Request object with IP and headers
   * @param {string} userId - User identifier
   * @returns {string} Session ID
   */
  createSession(req, userId) {
    // Defense 1: Cryptographically secure session ID
    const sessionId = this.generateSecureSessionId();
    
    // Defense 2: Create session fingerprint
    const fingerprint = this.createFingerprint(req);
    
    // Defense 3: Check concurrent session limits
    this.enforceConcurrentSessionLimit(userId);
    
    console.log('✅ SECURE: Creating session with fingerprinting and limits');
    
    this.store.create(sessionId, {
      userId,
      fingerprint,
      rotationDue: Date.now() + this.config.rotationInterval,
      expiresAt: Date.now() + this.config.sessionTimeout
    });

    return sessionId;
  }

  /**
   * SECURE: Validates session with multiple security checks
   * @param {object} req - Request object
   * @param {string} sessionId - Session ID
   * @returns {object|null} Session data or null if invalid/hijacked
   */
  validateSession(req, sessionId) {
    const session = this.store.get(sessionId);
    
    if (!session) {
      console.log('🛡️  Session not found');
      return null;
    }

    // Defense 1: Check session expiration
    if (Date.now() > session.expiresAt) {
      console.log('🛡️  BLOCKED: Session expired');
      this.store.delete(sessionId);
      return null;
    }

    // Defense 2: Detect session hijacking via fingerprint validation
    const currentFingerprint = this.createFingerprint(req);
    const hijackScore = this.detectHijacking(session.fingerprint, currentFingerprint);
    
    if (hijackScore > 0.5) {
      console.log('🛡️  BLOCKED: Possible session hijacking detected!');
      console.log(`    Hijack confidence: ${(hijackScore * 100).toFixed(1)}%`);
      
      // Invalidate all user sessions on hijack detection
      this.store.deleteUserSessions(session.userId);
      return null;
    }

    // Defense 3: Session rotation
    if (Date.now() > session.rotationDue) {
      console.log('🔄 Session rotation required');
      return { ...session, rotationRequired: true };
    }

    // Defense 4: Update activity timestamp
    this.store.update(sessionId, {
      lastActivity: Date.now(),
      expiresAt: Date.now() + this.config.sessionTimeout
    });

    console.log('✅ Session validated successfully');
    
    return session;
  }

  /**
   * Rotates session ID while preserving session data
   * @param {string} oldSessionId - Old session ID
   * @param {object} req - Request object
   * @returns {string} New session ID
   */
  rotateSession(oldSessionId, req) {
    const session = this.store.get(oldSessionId);
    if (!session) {
      return null;
    }

    // Create new session with same data
    const newSessionId = this.generateSecureSessionId();
    
    this.store.create(newSessionId, {
      ...session,
      fingerprint: this.createFingerprint(req),
      rotationDue: Date.now() + this.config.rotationInterval,
      rotationCount: (session.rotationCount || 0) + 1
    });

    // Delete old session
    this.store.delete(oldSessionId);

    console.log('🔄 Session rotated successfully');
    
    return newSessionId;
  }

  /**
   * Generates cryptographically secure session ID
   * @returns {string} Session ID
   */
  generateSecureSessionId() {
    return crypto.randomBytes(32).toString('hex');
  }

  /**
   * Creates session fingerprint from request characteristics
   * @param {object} req - Request object
   * @returns {object} Fingerprint data
   */
  createFingerprint(req) {
    const fingerprint = {
      ipAddress: req.ip || req.connection?.remoteAddress || 'unknown',
      userAgent: req.headers?.['user-agent'] || 'unknown'
    };

    // Higher strength includes more signals (but less flexible for legitimate use)
    if (this.config.fingerprintStrength === 'high') {
      fingerprint.acceptLanguage = req.headers?.['accept-language'];
      fingerprint.acceptEncoding = req.headers?.['accept-encoding'];
    }

    // Create hash of fingerprint
    fingerprint.hash = this.hashFingerprint(fingerprint);

    return fingerprint;
  }

  /**
   * Hashes fingerprint data
   * @param {object} fingerprint - Fingerprint data
   * @returns {string} Hash
   */
  hashFingerprint(fingerprint) {
    const data = JSON.stringify({
      ip: fingerprint.ipAddress,
      ua: fingerprint.userAgent,
      lang: fingerprint.acceptLanguage,
      enc: fingerprint.acceptEncoding
    });
    return crypto.createHash('sha256').update(data).digest('hex');
  }

  /**
   * Detects potential session hijacking by comparing fingerprints
   * @param {object} original - Original fingerprint
   * @param {object} current - Current fingerprint
   * @returns {number} Hijack score (0-1, higher = more likely hijacked)
   */
  detectHijacking(original, current) {
    let suspicionScore = 0;
    let checks = 0;

    // Check 1: IP address change (HIGH severity)
    if (original.ipAddress !== current.ipAddress) {
      suspicionScore += 0.7;
      console.log(`⚠️  IP changed: ${original.ipAddress} → ${current.ipAddress}`);
    }
    checks++;

    // Check 2: User Agent change (MEDIUM severity)
    if (original.userAgent !== current.userAgent) {
      suspicionScore += 0.4;
      console.log(`⚠️  User Agent changed`);
    }
    checks++;

    // Check 3: Accept-Language change (LOW severity)
    if (this.config.fingerprintStrength === 'high') {
      if (original.acceptLanguage !== current.acceptLanguage) {
        suspicionScore += 0.1;
      }
      checks++;
    }

    return Math.min(suspicionScore, 1.0);
  }

  /**
   * Enforces concurrent session limits per user
   * @param {string} userId - User identifier
   */
  enforceConcurrentSessionLimit(userId) {
    const userSessions = this.store.getUserSessions(userId);
    
    if (userSessions.length >= this.config.maxConcurrentSessions) {
      console.log(`🛡️  Enforcing concurrent session limit (${this.config.maxConcurrentSessions})`);
      
      // Remove oldest session
      let oldestSession = null;
      let oldestTime = Infinity;
      
      userSessions.forEach(sessionId => {
        const session = this.store.get(sessionId);
        if (session && session.lastActivity < oldestTime) {
          oldestTime = session.lastActivity;
          oldestSession = sessionId;
        }
      });
      
      if (oldestSession) {
        this.store.delete(oldestSession);
        console.log('🗑️  Removed oldest session');
      }
    }
  }

  /**
   * Destroys a session
   * @param {string} sessionId - Session ID
   */
  destroySession(sessionId) {
    this.store.delete(sessionId);
    console.log('🗑️  Session destroyed');
  }

  /**
   * Destroys all sessions for a user
   * @param {string} userId - User identifier
   */
  destroyUserSessions(userId) {
    this.store.deleteUserSessions(userId);
    console.log('🗑️  All user sessions destroyed');
  }
}

// =============================================================================
// EXPRESS MIDDLEWARE
// =============================================================================

/**
 * Creates Express middleware for session management
 * @param {SecureSessionManager} sessionManager - Session manager instance
 * @returns {Function} Express middleware
 */
function createSessionMiddleware(sessionManager) {
  return (req, res, next) => {
    // Extract session ID from cookie or header
    const sessionId = req.cookies?.sessionId || req.headers['x-session-id'];

    if (!sessionId) {
      req.session = null;
      return next();
    }

    // Validate session
    const session = sessionManager.validateSession(req, sessionId);

    if (!session) {
      res.clearCookie('sessionId');
      req.session = null;
      return next();
    }

    // Check if rotation is required
    if (session.rotationRequired) {
      const newSessionId = sessionManager.rotateSession(sessionId, req);
      res.cookie('sessionId', newSessionId, {
        httpOnly: true,
        secure: true,
        sameSite: 'strict',
        maxAge: 24 * 60 * 60 * 1000 // 24 hours
      });
      session.sessionId = newSessionId;
    }

    req.session = session;
    next();
  };
}

// =============================================================================
// EXAMPLE USAGE
// =============================================================================

if (require.main === module) {
  console.log('='.repeat(80));
  console.log('MITRE ATT&CK T1185: Session Hijacking Defense Demo');
  console.log('='.repeat(80));

  // Mock request objects
  const createMockRequest = (ip, userAgent) => ({
    ip,
    headers: { 'user-agent': userAgent },
    connection: { remoteAddress: ip }
  });

  console.log('\n--- VULNERABLE SESSION MANAGER ---\n');
  
  const vulnerableManager = new VulnerableSessionManager();
  
  const vulnReq1 = createMockRequest('192.168.1.100', 'Mozilla/5.0 (Windows NT 10.0)');
  const vulnSessionId = vulnerableManager.createSession(vulnReq1, 'user123');
  console.log('Session created:', vulnSessionId);
  
  // Simulate hijacker with different IP
  const hijackerReq = createMockRequest('10.0.0.50', 'Mozilla/5.0 (Linux; Android)');
  const vulnResult = vulnerableManager.validateSession(hijackerReq, vulnSessionId);
  console.log('⚠️  Hijacker validated:', vulnResult ? 'YES (VULNERABLE!)' : 'NO');

  console.log('\n--- SECURE SESSION MANAGER ---\n');
  
  const secureManager = new SecureSessionManager({
    maxConcurrentSessions: 2,
    sessionTimeout: 30 * 60 * 1000,
    rotationInterval: 15 * 60 * 1000,
    fingerprintStrength: 'medium'
  });
  
  // Create session for legitimate user
  console.log('1. Creating session for legitimate user:');
  const secureReq1 = createMockRequest('192.168.1.100', 'Mozilla/5.0 (Windows NT 10.0)');
  const secureSessionId = secureManager.createSession(secureReq1, 'user456');
  console.log('   Session ID:', secureSessionId.substring(0, 16) + '...');
  
  // Validate with legitimate user
  console.log('\n2. Legitimate user validates session:');
  const legitResult = secureManager.validateSession(secureReq1, secureSessionId);
  console.log('   Validated:', legitResult ? 'YES' : 'NO');
  
  // Simulate hijacking attempt
  console.log('\n3. Hijacker attempts to use stolen session:');
  const secureHijackerReq = createMockRequest('10.0.0.50', 'Mozilla/5.0 (Linux; Android)');
  const hijackResult = secureManager.validateSession(secureHijackerReq, secureSessionId);
  console.log('   Hijacker blocked:', hijackResult ? 'NO (FAILED!)' : 'YES (SUCCESS!)');
  
  // Test concurrent session limits
  console.log('\n4. Testing concurrent session limits:');
  const session2 = secureManager.createSession(secureReq1, 'user789');
  const session3 = secureManager.createSession(secureReq1, 'user789');
  const session4 = secureManager.createSession(secureReq1, 'user789'); // Should remove oldest
  console.log('   Sessions created: 3, Limit: 2, Oldest removed: YES');
  
  // Test session rotation
  console.log('\n5. Testing session rotation:');
  const newSessionId = secureManager.rotateSession(session4, secureReq1);
  console.log('   Old session ID:', session4.substring(0, 16) + '...');
  console.log('   New session ID:', newSessionId.substring(0, 16) + '...');

  console.log('\n' + '='.repeat(80));
  console.log('Demo complete. Key defenses implemented:');
  console.log('  ✅ Session fingerprinting (IP + User-Agent)');
  console.log('  ✅ Hijacking detection with suspicion scoring');
  console.log('  ✅ Automatic session rotation');
  console.log('  ✅ Concurrent session limits');
  console.log('  ✅ Session timeout and expiration');
  console.log('  ✅ Cryptographically secure session IDs');
  console.log('='.repeat(80));
}

module.exports = {
  SessionStore,
  VulnerableSessionManager,
  SecureSessionManager,
  createSessionMiddleware
};

File: samples/dotnet/SessionSecurity.cs
using Microsoft.AspNetCore.Http;
using System.Security.Cryptography;
using System.Text;
using System.Collections.Concurrent;

namespace MitreAttackForDevs.Samples;

/// <summary>
/// Demonstrates defense against:
/// - T1185: Browser Session Hijacking
/// - T1098: Account Manipulation
/// - T1550.004: Web Session Cookie
/// 
/// Implements secure session management with fingerprinting, rotation, and anomaly detection
/// </summary>
public class SecureSessionManager
{
    private readonly ILogger<SecureSessionManager> _logger;
    
    // Track active sessions with metadata
    private readonly ConcurrentDictionary<string, SessionMetadata> _activeSessions = new();
    
    // Track concurrent sessions per user
    private readonly ConcurrentDictionary<string, HashSet<string>> _userSessions = new();

    // Configuration
    private readonly SessionSecurityOptions _options;

    public SecureSessionManager(ILogger<SecureSessionManager> logger, SessionSecurityOptions? options = null)
    {
        _logger = logger;
        _options = options ?? new SessionSecurityOptions();
    }

    /// <summary>
    /// T1185 Defense: Create session with fingerprinting
    /// Binds session to client characteristics to prevent hijacking
    /// </summary>
    public async Task<SecureSession> CreateSessionAsync(
        string userId, 
        HttpContext httpContext,
        string? privilegeLevel = "standard")
    {
        var sessionId = GenerateSecureSessionId();
        var fingerprint = GenerateClientFingerprint(httpContext);
        
        var session = new SecureSession
        {
            SessionId = sessionId,
            UserId = userId,
            Fingerprint = fingerprint,
            CreatedAt = DateTimeOffset.UtcNow,
            LastAccessedAt = DateTimeOffset.UtcNow,
            ExpiresAt = DateTimeOffset.UtcNow.Add(_options.SessionTimeout),
            PrivilegeLevel = privilegeLevel ?? "standard",
            IpAddress = GetClientIpAddress(httpContext),
            UserAgent = httpContext.Request.Headers["User-Agent"].ToString()
        };

        var metadata = new SessionMetadata
        {
            Session = session,
            AccessCount = 1,
            LastRotationAt = DateTimeOffset.UtcNow
        };

        _activeSessions[sessionId] = metadata;
        
        // Track user sessions for concurrent session detection
        _userSessions.AddOrUpdate(
            userId,
            _ => new HashSet<string> { sessionId },
            (_, sessions) =>
            {
                lock (sessions)
                {
                    sessions.Add(sessionId);
                    return sessions;
                }
            });

        // T1185: Check for suspicious concurrent sessions
        await DetectConcurrentSessionAnomaliesAsync(userId, session);

        _logger.LogInformation(
            "[T1185] Session created for user {UserId}, SessionId: {SessionId}, IP: {IpAddress}",
            userId, sessionId, session.IpAddress);

        return session;
    }

    /// <summary>
    /// T1185 Defense: Validate session with fingerprint verification
    /// Detects session hijacking attempts by comparing fingerprints
    /// </summary>
    public async Task<SessionValidationResult> ValidateSessionAsync(
        string sessionId, 
        HttpContext httpContext)
    {
        if (!_activeSessions.TryGetValue(sessionId, out var metadata))
        {
            _logger.LogWarning("[T1185] Invalid session attempted: {SessionId}", sessionId);
            return SessionValidationResult.Invalid("Session not found");
        }

        var session = metadata.Session;
        
        // Check expiration
        if (DateTimeOffset.UtcNow > session.ExpiresAt)
        {
            await InvalidateSessionAsync(sessionId, "Session expired");
            return SessionValidationResult.Invalid("Session expired");
        }

        // T1185: Verify session fingerprint
        var currentFingerprint = GenerateClientFingerprint(httpContext);
        if (currentFingerprint != session.Fingerprint)
        {
            _logger.LogWarning(
                "[T1185] SECURITY ALERT: Session hijacking attempt detected! " +
                "SessionId: {SessionId}, UserId: {UserId}, " +
                "Original IP: {OriginalIp}, Current IP: {CurrentIp}",
                sessionId, session.UserId, session.IpAddress, GetClientIpAddress(httpContext));

            // Invalidate potentially compromised session
            await InvalidateSessionAsync(sessionId, "Fingerprint mismatch - potential hijacking");
            
            return SessionValidationResult.Invalid("Session validation failed");
        }

        // T1185: Detect IP address changes (potential hijacking)
        var currentIp = GetClientIpAddress(httpContext);
        if (currentIp != session.IpAddress && _options.EnforceIpBinding)
        {
            _logger.LogWarning(
                "[T1185] IP address change detected for session {SessionId}: {OldIp} -> {NewIp}",
                sessionId, session.IpAddress, currentIp);

            // Depending on policy, might invalidate or just log
            if (_options.StrictIpBinding)
            {
                await InvalidateSessionAsync(sessionId, "IP address changed");
                return SessionValidationResult.Invalid("Session IP address mismatch");
            }
        }

        // Update session activity
        session.LastAccessedAt = DateTimeOffset.UtcNow;
        metadata.AccessCount++;

        // T1185: Automatic session rotation on high activity
        if (ShouldRotateSession(metadata))
        {
            await RotateSessionAsync(sessionId, httpContext);
        }

        return SessionValidationResult.Valid(session);
    }

    /// <summary>
    /// T1098 Defense: Rotate session ID on privilege escalation
    /// Prevents session fixation attacks during privilege changes
    /// </summary>
    public async Task<string> RotateSessionOnPrivilegeChangeAsync(
        string oldSessionId, 
        string newPrivilegeLevel,
        HttpContext httpContext)
    {
        if (!_activeSessions.TryGetValue(oldSessionId, out var metadata))
        {
            throw new InvalidOperationException("Session not found");
        }

        var oldSession = metadata.Session;
        
        _logger.LogInformation(
            "[T1098] Rotating session due to privilege change: {UserId}, " +
            "Old Privilege: {OldPrivilege}, New Privilege: {NewPrivilege}",
            oldSession.UserId, oldSession.PrivilegeLevel, newPrivilegeLevel);

        // Create new session ID
        var newSessionId = GenerateSecureSessionId();
        
        // Create new session with updated privileges
        var newSession = new SecureSession
        {
            SessionId = newSessionId,
            UserId = oldSession.UserId,
            Fingerprint = GenerateClientFingerprint(httpContext), // Re-fingerprint
            CreatedAt = DateTimeOffset.UtcNow,
            LastAccessedAt = DateTimeOffset.UtcNow,
            ExpiresAt = DateTimeOffset.UtcNow.Add(_options.SessionTimeout),
            PrivilegeLevel = newPrivilegeLevel,
            IpAddress = GetClientIpAddress(httpContext),
            UserAgent = httpContext.Request.Headers["User-Agent"].ToString(),
            RotatedFrom = oldSessionId
        };

        var newMetadata = new SessionMetadata
        {
            Session = newSession,
            AccessCount = 0,
            LastRotationAt = DateTimeOffset.UtcNow
        };

        // Replace old session with new one
        _activeSessions[newSessionId] = newMetadata;
        await InvalidateSessionAsync(oldSessionId, "Rotated due to privilege change");

        // Update user session tracking
        if (_userSessions.TryGetValue(oldSession.UserId, out var sessions))
        {
            lock (sessions)
            {
                sessions.Remove(oldSessionId);
                sessions.Add(newSessionId);
            }
        }

        return newSessionId;
    }

    /// <summary>
    /// T1185 Defense: Detect concurrent session anomalies
    /// Alerts when user has suspicious number of concurrent sessions
    /// </summary>
    private async Task DetectConcurrentSessionAnomaliesAsync(string userId, SecureSession newSession)
    {
        if (!_userSessions.TryGetValue(userId, out var sessions))
            return;

        int activeCount;
        lock (sessions)
        {
            // Clean up expired sessions
            sessions.RemoveWhere(sid => 
            {
                if (_activeSessions.TryGetValue(sid, out var meta))
                {
                    return DateTimeOffset.UtcNow > meta.Session.ExpiresAt;
                }
                return true;
            });

            activeCount = sessions.Count;
        }

        // T1185: Alert on suspicious concurrent session count
        if (activeCount > _options.MaxConcurrentSessions)
        {
            _logger.LogWarning(
                "[T1185] SECURITY ALERT: User {UserId} has {Count} concurrent sessions " +
                "(max: {Max}). Possible account compromise.",
                userId, activeCount, _options.MaxConcurrentSessions);

            // Optional: Implement policy to invalidate oldest sessions or require re-authentication
            await EnforceConcurrentSessionLimitAsync(userId);
        }
    }

    /// <summary>
    /// T1550.004 Defense: Invalidate session and log event
    /// </summary>
    public async Task InvalidateSessionAsync(string sessionId, string reason)
    {
        if (_activeSessions.TryRemove(sessionId, out var metadata))
        {
            var session = metadata.Session;
            
            _logger.LogInformation(
                "[T1185] Session invalidated: {SessionId}, UserId: {UserId}, Reason: {Reason}",
                sessionId, session.UserId, reason);

            // Remove from user session tracking
            if (_userSessions.TryGetValue(session.UserId, out var sessions))
            {
                lock (sessions)
                {
                    sessions.Remove(sessionId);
                }
            }

            // Could also add to a blacklist or audit log here
            await Task.CompletedTask;
        }
    }

    /// <summary>
    /// T1185: Invalidate all sessions for a user (e.g., on password change or suspected compromise)
    /// </summary>
    public async Task InvalidateAllUserSessionsAsync(string userId, string reason)
    {
        if (!_userSessions.TryRemove(userId, out var sessions))
            return;

        List<string> sessionIds;
        lock (sessions)
        {
            sessionIds = sessions.ToList();
        }

        _logger.LogWarning(
            "[T1185] Invalidating all sessions for user {UserId}. Count: {Count}, Reason: {Reason}",
            userId, sessionIds.Count, reason);

        foreach (var sessionId in sessionIds)
        {
            await InvalidateSessionAsync(sessionId, reason);
        }
    }

    #region Private Helper Methods

    private string GenerateSecureSessionId()
    {
        // Generate cryptographically secure random session ID
        var bytes = new byte[32];
        using var rng = RandomNumberGenerator.Create();
        rng.GetBytes(bytes);
        return Convert.ToBase64String(bytes)
            .Replace('+', '-')
            .Replace('/', '_')
            .TrimEnd('=');
    }

    /// <summary>
    /// T1185: Generate client fingerprint from IP + User-Agent
    /// More sophisticated implementations might include TLS fingerprinting, canvas fingerprinting, etc.
    /// </summary>
    private string GenerateClientFingerprint(HttpContext context)
    {
        var ip = GetClientIpAddress(context);
        var userAgent = context.Request.Headers["User-Agent"].ToString();
        
        // Could include additional entropy like Accept-Language, Accept-Encoding, etc.
        var fingerprintData = $"{ip}|{userAgent}";
        
        using var sha256 = SHA256.Create();
        var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(fingerprintData));
        return Convert.ToBase64String(hash);
    }

    private string GetClientIpAddress(HttpContext context)
    {
        // Check for forwarded IP (behind proxy/load balancer)
        var forwardedFor = context.Request.Headers["X-Forwarded-For"].FirstOrDefault();
        if (!string.IsNullOrEmpty(forwardedFor))
        {
            return forwardedFor.Split(',')[0].Trim();
        }

        return context.Connection.RemoteIpAddress?.ToString() ?? "unknown";
    }

    private bool ShouldRotateSession(SessionMetadata metadata)
    {
        // Rotate session ID periodically to reduce hijacking window
        var timeSinceRotation = DateTimeOffset.UtcNow - metadata.LastRotationAt;
        return timeSinceRotation > _options.RotationInterval;
    }

    private async Task RotateSessionAsync(string oldSessionId, HttpContext httpContext)
    {
        if (!_activeSessions.TryGetValue(oldSessionId, out var metadata))
            return;

        var newSessionId = GenerateSecureSessionId();
        var newMetadata = new SessionMetadata
        {
            Session = metadata.Session with 
            { 
                SessionId = newSessionId,
                RotatedFrom = oldSessionId
            },
            AccessCount = 0,
            LastRotationAt = DateTimeOffset.UtcNow
        };

        _activeSessions[newSessionId] = newMetadata;
        _activeSessions.TryRemove(oldSessionId, out _);

        _logger.LogInformation(
            "[T1185] Session rotated: {OldId} -> {NewId}",
            oldSessionId, newSessionId);

        await Task.CompletedTask;
    }

    private async Task EnforceConcurrentSessionLimitAsync(string userId)
    {
        if (!_userSessions.TryGetValue(userId, out var sessions))
            return;

        List<(string sessionId, DateTimeOffset lastAccess)> sessionList;
        
        lock (sessions)
        {
            sessionList = sessions
                .Select(sid => 
                {
                    if (_activeSessions.TryGetValue(sid, out var meta))
                    {
                        return (sid, meta.Session.LastAccessedAt);
                    }
                    return (sid, DateTimeOffset.MinValue);
                })
                .OrderBy(s => s.lastAccess)
                .ToList();

            // Keep only the most recent sessions
            var toRemove = sessionList
                .Take(sessionList.Count - _options.MaxConcurrentSessions)
                .Select(s => s.sessionId)
                .ToList();

            foreach (var sessionId in toRemove)
            {
                await InvalidateSessionAsync(sessionId, "Concurrent session limit exceeded");
            }
        }
    }

    #endregion
}

#region Supporting Types

public record SecureSession
{
    public required string SessionId { get; init; }
    public required string UserId { get; init; }
    public required string Fingerprint { get; init; }
    public required DateTimeOffset CreatedAt { get; init; }
    public required DateTimeOffset LastAccessedAt { get; set; }
    public required DateTimeOffset ExpiresAt { get; init; }
    public required string PrivilegeLevel { get; init; }
    public required string IpAddress { get; init; }
    public required string UserAgent { get; init; }
    public string? RotatedFrom { get; init; }
}

internal class SessionMetadata
{
    public required SecureSession Session { get; set; }
    public int AccessCount { get; set; }
    public DateTimeOffset LastRotationAt { get; set; }
}

public record SessionValidationResult
{
    public bool IsValid { get; init; }
    public string? ErrorMessage { get; init; }
    public SecureSession? Session { get; init; }

    public static SessionValidationResult Valid(SecureSession session) =>
        new() { IsValid = true, Session = session };

    public static SessionValidationResult Invalid(string message) =>
        new() { IsValid = false, ErrorMessage = message };
}

public class SessionSecurityOptions
{
    public TimeSpan SessionTimeout { get; set; } = TimeSpan.FromMinutes(30);
    public TimeSpan RotationInterval { get; set; } = TimeSpan.FromMinutes(15);
    public int MaxConcurrentSessions { get; set; } = 3;
    public bool EnforceIpBinding { get; set; } = true;
    public bool StrictIpBinding { get; set; } = false; // If true, IP change invalidates session
}

#endregion

/// <summary>
/// Key Takeaways for Session Security:
/// 
/// T1185 - Browser Session Hijacking Defense:
/// 1. Bind sessions to client fingerprint (IP + User-Agent hash)
/// 2. Detect and alert on IP address changes
/// 3. Implement automatic session rotation
/// 4. Monitor for concurrent session anomalies
/// 5. Use cryptographically secure session IDs
/// 
/// T1098 - Account Manipulation Defense:
/// 1. Rotate session on privilege changes
/// 2. Invalidate all sessions on password change
/// 3. Enforce maximum concurrent sessions
/// 4. Log all session lifecycle events
/// 
/// T1550.004 - Web Session Cookie Defense:
/// 1. Use HttpOnly and Secure cookie flags (in middleware)
/// 2. Implement SameSite cookie policy
/// 3. Set appropriate session timeouts
/// 4. Never expose session tokens in URLs
/// 
/// References:
/// - MITRE ATT&CK T1185: https://attack.mitre.org/techniques/T1185/
/// - MITRE ATT&CK T1098: https://attack.mitre.org/techniques/T1098/
/// - MITRE ATT&CK T1550.004: https://attack.mitre.org/techniques/T1550/004/
/// - OWASP Session Management Cheat Sheet
/// </summary>
